### PR TITLE
Docker images for Stargate v2

### PR DESCRIPTION
**Pull request by jeffreyscarpenter**
Build Docker images for Stargate v2 in Stargate repo to make it simpler to use images for testing during development

### CODE REVIEW

File: Dockerfile
Comment (tatu-at-datastax) [inline Dockerfile:32]: Would this still leave those jars as part of the image (or rather, use up space)? I recall Docker image construction is sometimes tricky as snapshot is taken between steps.

File: docker-compose/cassandra-3.11/docker-compose.yml
Comment (tatu-at-datastax) [inline docker-compose/cassandra-3.11/docker-compose.yml:4]: I am not familir with Docker compose, so.... is this naming convention that must be followed? Or can we use more meaningful names? (I assume former)

Comment (tatu-at-datastax) [reply docker-compose/cassandra-3.11/docker-compose.yml:4]: Oh actually it looks like this can be freely selected (looking at later entries). So would it make sense to call it "cassandra-1" or "storage-1"?

**Review by tatu-at-datastax — APPROVED** [review]

File: docker-compose/cassandra-3.11/docker-compose.yml
Comment (tatu-at-datastax) [reply docker-compose/cassandra-3.11/docker-compose.yml:4]: I assume this mostly just follows existing naming convention within Datastax tho. If so, perhaps that makes sense.

Comment (jeffreyscarpenter) [reply docker-compose/cassandra-3.11/docker-compose.yml:4]: I preserved the previous conventions from the copies in [docker images](https://github.com/stargate/docker-images), but agree this could be more descriptive.

File: Dockerfile
Comment (tatu-at-datastax) [inline Dockerfile:28]: Wouldn't this copy all jars, including persistence ones?

File: docker-compose/cassandra-3.11/start_cass_311.sh
Comment (tatu-at-datastax) [inline docker-compose/cassandra-3.11/start_cass_311.sh:1]: Newbie question: does this refer to `docker-compose.yml` via naming convention (defaulting)? Or how is the dependency indicated?

File: docker-compose/cassandra-4.0/README.md
Comment (tatu-at-datastax) [inline docker-compose/cassandra-4.0/README.md:15]: Ahh. I think this answers my earlier question; there's default name, but one that can be overridden.

**Comment by ivansenic** [comment]
@jeffreyscarpenter I would like to invest time in reviewing this, but would need more context here.. So what are we trying to achive, are all the files newly written or was there some starting point, is all of this staying after in v2 or is something gonna become obsolete? Thanks in advance..

**Comment by jeffreyscarpenter** [comment]
> @jeffreyscarpenter I would like to invest time in reviewing this, but would need more context here.. So what are we trying to achive, are all the files newly written or was there some starting point, is all of this staying after in v2 or is something gonna become obsolete? Thanks in advance..

Context:
One of the overall goals for v2 is to be more Docker/Kubernetes friendly. This PR represents one step in that direction, to to build Docker images in the main stargate repo as part of the standard build process so they can be more easily used during development. (This is different than the current process as implemented via the `docker-images` repo, which requires a Stargate release.) Other subsequent PRs will add actions to build, tag and publish these Docker images as part of our release process, and provide options to run some tests using these images using Docker and Kubernetes. It is likely that we'll want to deprecate the `docker-images` repo when v2 becomes the main version, but we don't have to make a final decision on that now.

Details:
The Dockerfile creates images that are similar to those created by `docker-images`, although they use a more recent JVM and are more streamlined. A single Dockerfile is used for simplicity and so we can have a common base image for the different coordinator images. The coordinator images have names with prefix `coordinator-` instead of `stargate-`. This provides a more accurate name of what these images contain and will help avoid confusion with v1 images. The fact that images will be tagged `v2.0.0-ALPHA1` (and following) when published will also help avoid unintentional use of the wrong images by current v1 users.

The Dockerfile also contains instructions for building the REST API image, although this could easily be broken out into a separate Dockerfile, for example if we decide to move service implementations to their own GitHub repos at some point in the future. You'll note that it uses Java 11.

The `docker-compose` files are largely based on the examples from `docker-images` with updates to have consistent naming. There are also new compose files that support running coordinators in "developer mode" that can be used for quick startup without having to start a Cassandra cluster. I've already found these pretty useful for local testing.

File: Dockerfile
Comment (ivansenic) [inline Dockerfile:48]: is there anything that stops us from using java 17 here?

File: Dockerfile
Comment (ivansenic) [reply Dockerfile:28]: Seems to me as well.. Ahh OK, seen the trick in the build file.. Why not simply going for:

```
RUN rm /stargate-lib/persistence*.jar
```

File: build_docker_images.sh
Comment (ivansenic) [inline build_docker_images.sh:6]: I think we should always point to a wrapper and skip tests:

`./mvnw install -P dse -DskipTests=true`

File: build_docker_images.sh
Comment (ivansenic) [inline build_docker_images.sh:19]: I guess this can be omitted if you apply the above proposed `RUN rm ..`?

File: docker-compose/cassandra-3.11/.env
Comment (ivansenic) [inline docker-compose/cassandra-3.11/.env:2]: Can you extend the `README.md` in the `persistence-cassandra-3.11` and add the update of this value to the list of tasks we need to do when upgrading the version?

File: docker-compose/cassandra-3.11/docker-compose.yml
Comment (ivansenic) [inline docker-compose/cassandra-3.11/docker-compose.yml:14]: hmm is this correct, point to itself as the seed?

File: docker-compose/cassandra-3.11/docker-compose.yml
Comment (ivansenic) [inline docker-compose/cassandra-3.11/docker-compose.yml:39]: Not sure why we have one coordinator and three C* nodes, but would it actually make more sense to have two each? Also not sure if we can use `replicas` here, at least for the coordinator?

File: docker-compose/cassandra-4.0/.env
Comment (ivansenic) [inline docker-compose/cassandra-4.0/.env:2]: should we also fix this to the version we currently support `4.0.1`?

File: docker-compose/dse-6.8/.env
Comment (ivansenic) [inline docker-compose/dse-6.8/.env:2]: we use `6.8.16`

File: kubernetes/k8ssandra-values.yaml
Comment (ivansenic) [inline kubernetes/k8ssandra-values.yaml:2]: 4.0.1?

File: kubernetes/k8ssandra-values.yaml
Comment (ivansenic) [inline kubernetes/k8ssandra-values.yaml:9]: well this seems odd? new gen size matches the whole heap?

File: kubernetes/k8ssandra-values.yaml
Comment (ivansenic) [inline kubernetes/k8ssandra-values.yaml:30]: should we push this a bit up?

**Review by ivansenic — CHANGES_REQUESTED** [review]
Does not compile, probably needs rebasing, can not test the scripts.. I added some other comments as well, have a look..

I guess that the restapi image would become obsolete once we move to new framework, but for now it's good.. Really good work @jeffreyscarpenter :+1:

File: Dockerfile
Comment (jeffreyscarpenter) [reply Dockerfile:28]: @ivansenic I originally had a step in the image creation process that copied in all of the jar files and then removed the undesired ones but @tatu-at-datastax pointed out that this would be unnecessarily bloat the image since the files are still technically there but just don't appear in the directory. Plus it s a bit more complex than removing all `persistence*.jar` files, there are actually ones you want to keep per image.

File: Dockerfile
Comment (jeffreyscarpenter) [reply Dockerfile:48]: probably not, but I suggest we wait on this until we select the updated Java framework, which may make this section of the Dockerfile obsolete.

File: kubernetes/k8ssandra-values.yaml
Comment (jeffreyscarpenter) [reply kubernetes/k8ssandra-values.yaml:2]: I'm going to pull the Kubernetes config from this PR, it is still a WIP that I included inadvertently.

File: docker-compose/cassandra-3.11/start_cass_311.sh
Comment (jeffreyscarpenter) [reply docker-compose/cassandra-3.11/start_cass_311.sh:1]: yes, that's correct. `docker-compose.yml` is the default name.

File: build_docker_images.sh
Comment (jeffreyscarpenter) [reply build_docker_images.sh:6]: ok

File: docker-compose/cassandra-3.11/docker-compose.yml
Comment (jeffreyscarpenter) [reply docker-compose/cassandra-3.11/docker-compose.yml:14]: confirmed this is not necessary, removing

File: docker-compose/dse-6.8/.env
Comment (jeffreyscarpenter) [reply docker-compose/dse-6.8/.env:2]: fixing

File: docker-compose/cassandra-4.0/.env
Comment (jeffreyscarpenter) [reply docker-compose/cassandra-4.0/.env:2]: fixing

File: docker-compose/cassandra-3.11/.env
Comment (jeffreyscarpenter) [reply docker-compose/cassandra-3.11/.env:2]: fixing

File: docker-compose/cassandra-3.11/README.md
Comment (ivansenic) [inline docker-compose/cassandra-3.11/README.md:15]: all `docker-compose` files have a `.yml` extension and not `.yaml`.. Either docs or files should be adapted..

**Review by ivansenic — CHANGES_REQUESTED** [review]
The branch is still not compiling:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project testing: Compilation failure: Compilation failure: 
[ERROR] /home/ise/workspace/datastax/stargate/stargate/testing/src/main/java/io/stargate/it/grpc/SchemaTest.java:[45,13] cannot find symbol
[ERROR]   symbol:   method describeKeyspace(io.stargate.proto.Schema.DescribeKeyspaceQuery)
[ERROR]   location: variable stub of type io.stargate.proto.StargateGrpc.StargateBlockingStub
[ERROR] /home/ise/workspace/datastax/stargate/stargate/testing/src/main/java/io/stargate/it/grpc/SchemaTest.java:[64,70] cannot find symbol
[ERROR]   symbol:   method getName()
[ERROR]   location: class io.stargate.proto.QueryOuterClass.TypeSpec.Udt
[ERROR] /home/ise/workspace/datastax/stargate/stargate/testing/src/main/java/io/stargate/it/grpc/SchemaTest.java:[68,36] cannot find symbol
[ERROR]   symbol:   method getName()
[ERROR]   location: class io.stargate.proto.QueryOuterClass.TypeSpec.Udt
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :testing
```

I run the `./build_docker_images.sh` and nothing is being built.. I get:
```
$ ./build_docker_images.sh 
cp: -r not specified; omitting directory './stargate-lib/felix-cache'
cp: -r not specified; omitting directory './stargate-lib/log'
```

So I don't get it, is this working for you? If so then there is some issue with the mac/linux shell compatibility.. Let me know and maybe I can dig deeper?

File: docker-compose/cassandra-4.0/start_cass_40.sh
Comment (ivansenic) [inline docker-compose/cassandra-4.0/start_cass_40.sh:32]:
```suggestion
while [ "$(curl -s -o /dev/null -w ''%{http_code}'' http://localhost:8082/health)" != "200" ]; do
    printf '.'
    sleep 5
done
```

**Review by ivansenic — APPROVED** [review]
I am approving, but there are 4 remaining comments and some suggestions to be solved please :pray:

File: docker-compose/cassandra-4.0/start_cass_40.sh
Comment (jeffreyscarpenter) [reply docker-compose/cassandra-4.0/start_cass_40.sh]: fixed

File: docker-compose/cassandra-3.11/start_cass_311.sh
Comment (jeffreyscarpenter) [reply docker-compose/cassandra-3.11/start_cass_311.sh]: fixed

File: docker-compose/dse-6.8/start_dse_68.sh
Comment (jeffreyscarpenter) [reply docker-compose/dse-6.8/start_dse_68.sh]: fixed

File: Dockerfile
Comment (jeffreyscarpenter) [reply Dockerfile]: agree, also changing persistence base image to `openjdk:8-jre-slim`, this will give us latest JVMs. We do not need strict reproducibility of Docker image builds.

File: docker-compose/cassandra-3.11/docker-compose.yml
Comment (jeffreyscarpenter) [reply docker-compose/cassandra-3.11/docker-compose.yml:66]: fixed, this port expose command was left in by mistake.

File: build_docker_images.sh
Comment (jeffreyscarpenter) [reply build_docker_images.sh]: added clean to the sample `mvnw` command

File: docker-compose/cassandra-3.11/README.md
Comment (jeffreyscarpenter) [reply docker-compose/cassandra-3.11/README.md]: updated readme files.

**Comment by tatu-at-datastax** [comment]
The only additional comment/question I have is about use of `latest` for JDK: I can see benefits of always using the latest, as 
JDK maintainers have very good tracker with compatibility (regressions rare). But at the same time it is usually a good practice to know exact version that is being packaged.

So: do we think in general that we should just use the latest patch for given JDK major version, trusting `latest` tagging?

**Comment by ivansenic** [comment]
> The only additional comment/question I have is about use of `latest` for JDK: I can see benefits of always using the latest, as JDK maintainers have very good tracker with compatibility (regressions rare). But at the same time it is usually a good practice to know exact version that is being packaged.
> 
> So: do we think in general that we should just use the latest patch for given JDK major version, trusting `latest` tagging?

I was the one voting for the latest.. It's only to get the Java patches and security improvements as fast as possible.. Otherwise, you kinda forget to update the image base.. Same is already happening for us in the `docker-images` repo, we rarely update the base version.

**Comment by tatu-at-datastax** [comment]
@ivansenic I agree that it's very easy to just never update anything. And I personally have never observed an actual regression by latest JDK patches. So I'm fine with this. I just wanted to ask if anyone felt differently since I know there are developers who consider reproducible build (and explicit versions) important.

So, for JDKs, let's just use the latest patch for major version as suggested.

**Pull request merged**
